File: python3-bottle/rpn.py
from collections import deque

BINARY_OPERATORS = {
    '+': lambda op1, op2: op2 + op1,
    '-': lambda op1, op2: op2 - op1,
    '*': lambda op1, op2: op2 * op1,
    '/': lambda op1, op2: op2 / op1,
    '^': lambda op1, op2: op2 ** op1
}

class RpnException(Exception):
    pass

def evaluate(tokens):
    stack = deque()
    for token in tokens:
        if token.isdigit():
            stack.append(int(token))
        elif token in BINARY_OPERATORS:
            if len(stack) < 2:
                raise RpnException('Stack underflow')
            stack.append(
                BINARY_OPERATORS[token](
                    stack.pop(),
                    stack.pop()
                )
            )
        else:
            raise RpnException('Unknown token')

    if len(stack) != 1:
        raise RpnException('Multiple stack elements remain')

    return stack[0]

File: golang/src/github.com/DanielBrookRoberge/rpn/rpn.go
package rpn

import "math"
import "strconv"

type binopt func(float64, float64) float64

var binaryOperators = map[string]binopt{
	"+": func(op1, op2 float64) float64 { return op2 + op1 },
	"-": func(op1, op2 float64) float64 { return op2 - op1 },
	"*": func(op1, op2 float64) float64 { return op2 * op1 },
	"/": func(op1, op2 float64) float64 { return op2 / op1 },
	"^": func(op1, op2 float64) float64 { return math.Pow(op2, op1) },
}

func Evaluate(tokens []string) float64 {
	stack := make([]float64, 0);

	for _, token := range tokens {
		if num, err := strconv.ParseFloat(token, 64); err == nil {
			stack = append(stack, num)
		} else if operation, ok := binaryOperators[token]; ok {
			// This will panic if there is a stack underflow
			op1 := stack[len(stack)-1]
			op2 := stack[len(stack)-2]
			stack = stack[:len(stack)-2]
			stack = append(stack, operation(op1, op2))
		} else {
			panic("Unknown token")
		}
	}
	if len(stack) != 1 {
		panic("Multiple values left on stack")
	}
	return stack[0]
}
